### PR TITLE
Ensure genesis state retrieval doesn't use zero hashes root

### PR DESCRIPTION
**Pull request by terencechain**
**What type of PR is this?**

> Bug fix

**What does this PR do? Why is it needed?**

Discovered in Spadina, without finality and restarting node will trigger beacon node to retrieve state using zero hashes. This fixed it and added a regression test

### CODE REVIEW

**Review by rauljordan — DISMISSED** [review]

File: beacon-chain/blockchain/service.go [hunk 1]
```diff
@@ -146,9 +146,22 @@ func (s *Service) Start() {
...
+			if genesisBlock != nil {
+				r, err = genesisBlock.Block.HashTreeRoot()
+				if err != nil {
+					log.Fatalf("Could not tree hash genesis block: %v", err)
+				}
+			}
+			beaconState, err = s.stateGen.StateByRoot(s.ctx, r)
+			if err != nil {
+				log.Fatalf("Could not fetch beacon state by root: %v", err)
+			}
```
Comment (rkapka) [inline beacon-chain/blockchain/service.go:164]: Shouldn't this part be inside an `else` statement?

File: beacon-chain/blockchain/service.go [hunk 2]
```diff
@@ -146,7 +146,20 @@ func (s *Service) Start() {
 	}
 
 	if beaconState == nil {
-		beaconState, err = s.stateGen.StateByRoot(s.ctx, bytesutil.ToBytes32(cp.Root))
+		r := bytesutil.ToBytes32(cp.Root)
+		if r == params.BeaconConfig().ZeroHash {
```
Comment (nisdas) [inline beacon-chain/blockchain/service.go:150]: Add a comment here on why we check for zero hashes here.

File: beacon-chain/blockchain/service.go [hunk 3]
```diff
@@ -146,7 +146,23 @@ func (s *Service) Start() {
...
+		r := bytesutil.ToBytes32(cp.Root)
+		// Before the first finalized epoch, in the current epoch,
+		// the finalized root is defined as zero hashes instead of genesis root hash.
+		// We want to use genesis root to retrieve for state.
+		if r == params.BeaconConfig().ZeroHash {
+			genesisBlock, err := s.beaconDB.GenesisBlock(s.ctx)
+			if err != nil {
+				log.Fatalf("Could not fetch finalized cp: %v", err)
+			}
+			if genesisBlock != nil {
```
Comment (nisdas) [inline beacon-chain/blockchain/service.go:158]: if genesis block is nil we will still end up doing `StateByRoot` , which I am assuming isnt the desired behaviour.

Comment (terencechain) [reply beacon-chain/blockchain/service.go:158]: That's fine. We want the beacon state to be `nil` then so we can enter `Waiting to reach the validator deposit threshold to start the beacon chain...` code block.

I've tested this case as well

**Review by nisdas — APPROVED** [review]

**Pull request merged**